Enhance replicate delete policy

gerrod3 opened the pull request:
Added a new field `delete_policy` that will determine how replicate will clean up the extra/leftover objects present in the downstream Pulp. The field is set to the default `all` which is the current behavior of deleting everything. The other two options are `none` and `labeled`. `none` is self-explanatory. Replicate now also labels each object created/updated on a replicate with the label `UpstreamPulp:{upstream.pk}`. This new label is what is used by the `labeled` policy to determine which objects should be deleted. Only objects with this label and are not present in the Upstream will be deleted under this policy, everything else will be left.

Also, added copying the Upstream objects' labels onto the downstream. This change isn't strictly necessary for the new delete feature, so I can remove it, add a new field for customization, or move it into another PR if wanted. 


fixes: #5214

mdellweg commented:
Please take the following as a pure observation, something to be aware of:
This is (or in case of acceptance would be) the first time a Pulp feature depends on labels stored on it's own entities.
Yes, replicate has relied on remote labels before. But that's a different thing. Labels have been there to bridge gaps for integrators and are in full control of the user.
I can not state this is good or bad it just feels a bit odd.

mdellweg commented on pulpcore/app/models/replica.py:
```diff
@@ -11,6 +11,15 @@
+    NONE = "none"
+    DELETE_CHOICES = (
+        (ALL, "Delete all local objects not present on upstream."),
+        (LABELED, "Delete local labeled objects not present on upstream."),
```
What should happen if you have a local repository without a label. Should replicate replace that, leave it alone or simply fail?

gerrod3 replied:
Note that my description isn't that great. The 'local labeled objects' refers to objects that have been labeled by replicate with `UpstreamPulp:{id}`. So any repository without this label means it wasn't created(tracked) by replicate and thus will not be removed under this policy. Say there is a unlabeled repository with the same name as an upstream one (a case our users will be in on their first replicate if this is merged), then that repository will get the new label and then be tracked for all future replicates.

mdellweg replied:
Will we set the labels anyway? Then users could run one more "old style" replication to label all the affected ones and change the flag then.
Now we could choose to "protect" manually created repositories from the replicators.

gerrod3 commented:
Ok I updated the PR to change the field to just `policy`. It does almost the exact same behavior, but now it also doesn't update existing local objects during the `LABELED` policy if there is a name collision and the object does not have a matching UpstreamPulp label. Also, renamed the `NONE` policy to `NODELETE`.

mdellweg reviewed (commented):
Damn, this review was pending for some days...

mdellweg commented on pulpcore/app/replica.py:
```diff
@@ -77,6 +78,13 @@ def needs_update(fields_dict, model_instance):
 
+    def _is_managed(self, object):
+        """Checks to see if the passed in object should be managed by replicate."""
+        if self.server.policy == UpstreamPulp.LABELED:
+            if object.pulp_labels.get("UpstreamPulp") != str(self.server.pk):
+                return False
+        return True
```
I'm wondering if that would make a forth policy. We can also just document that and maybe add the forth policy later.

mdellweg commented:
There's a migration conflict needing this to rebase.
